### ISSUE TITLE
ImageAugmentation using tf.keras.preprocessing.image.ImageDataGenerator and tf.datasets: model.fit() is running infinitely

### ISSUE THREAD

**Issue by Thomas-K-John** [tensorflow/tensorflow#39277]
**What I need help with / What I was wondering**
I am facing issue while running the fit() function in TensorFlow(v 2.2.0-rc4) with augmented images(using ImageDataGenerator) passed as a dataset. The fit() function is running infinitely without stopping.

**What I've tried so far**
I tried it with the default code which was shared in Tensorflow documentation. 

Please find the code snippet below:

import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import Sequential, Model 
from tensorflow.keras.layers import Dense, Dropout, Flatten
from tensorflow.keras.layers import Conv2D, MaxPooling2D 
from tensorflow.keras.layers import Input, Dense

flowers = tf.keras.utils.get_file(
    'flower_photos',
    'https://storage.googleapis.com/download.tensorflow.org/example_images/flower_photos.tgz',
    untar=True)

img_gen = tf.keras.preprocessing.image.ImageDataGenerator(rescale=1./255, rotation_range=20)

images, labels = next(img_gen.flow_from_directory(flowers))

print(images.dtype, images.shape)
print(labels.dtype, labels.shape)

train_data_gen = img_gen.flow_from_directory(
                    batch_size=32, 
                    directory=flowers,
                    shuffle=True,
                    target_size=(256, 256),
                    class_mode='categorical')

ds = tf.data.Dataset.from_generator(lambda: train_data_gen,
                     output_types=(tf.float32, tf.float32),
                     output_shapes=([32, 256, 256, 3],
                                    [32, 5])
                     )

ds = ds.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)

it = iter(ds)
batch = next(it)
print(batch)

def create_model():
  model = Sequential()
  model.add(Conv2D(32, (3, 3), activation='relu', input_shape=images[0].shape))
  model.add(Conv2D(32, (3, 3), activation='relu'))
  model.add(MaxPooling2D(pool_size=(2, 2)))
  model.add(Dropout(0.5))
  model.add(Conv2D(64, (3, 3), activation='relu'))
  model.add(Conv2D(64, (3, 3), activation='relu'))
  model.add(MaxPooling2D(pool_size=(2, 2)))
  model.add(Dropout(0.5))
  model.add(Flatten())
  model.add(Dense(64, activation='relu'))
  model.add(Dropout(0.5))
  model.add(Dense(5, activation='softmax'))
  return model

model = create_model()
model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=["accuracy"])
model.fit(ds, verbose=1,  batch_size= 32, epochs =1)

This last line of code - fit() is running infinitly without stopping. I had also tried passing steps_per_epoch = total_no_of_train_records/batch_size.

**It would be nice if...**
I would like you to confirm whethere this is a bug in the tensorflow datasets package and in which release will this be fixed.

**Environment information**
* System: Google colaborator
* Python version: v3.6.9
* `tensorflow version: v2.2.0-rc4

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.2-rc4 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/186467b0cc451d9463b49d5b866a4c70/untitled868.ipynb).Thanks!

**Comment by gowthamkpr** [comment]
I confirm this. Its running forever

**Comment by tomerk** [comment]
The ImageDataGenerator returns an infinite number of values, so the epoch would never end unless you specify `steps_per_epoch`.
The examples in ImageDataGenerator briefly mention this, but I think the documentation should be clearer about this. So I'll go ahead and update that.
https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/ImageDataGenerator

-----------
I'm not sure why setting `steps_per_epoch` didn't work for you? When I set a value for steps_per_epoch in this colab it ends once it has completed that many steps. Perhaps it's just being slow?
There are ~114 steps in one epoch over the dataset (3670 training images / 32 batch size). 114 * 7 sec / step (in this colab) is close to 15 minutes of runtime.

**Comment by Thomas-K-John** [comment]
I tried to use the model.fit() method by passing the steps_per_epoch=3670/32 and I got the following error:
**Code used to run the fit() method used:**
model.fit(ds, verbose=1,  steps_per_epoch=3670/32, epochs =1)
**Error:**
112/114 [============================>.] - ETA: 25s - loss: 1.9140 - accuracy: 0.3262
---------------------------------------------------------------------------
InvalidArgumentError                      Traceback (most recent call last)
<ipython-input-13-e56f4562ba98> in <module>()
----> 1 model.fit(ds, verbose=1,  steps_per_epoch=3670/32, epochs =1)

8 frames
/usr/local/lib/python3.6/dist-packages/tensorflow/python/eager/execute.py in quick_execute(op_name, num_outputs, inputs, attrs, ctx, name)
     58     ctx.ensure_initialized()
     59     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,
---> 60                                         inputs, attrs, num_outputs)
     61   except core._NotOkStatusException as e:
     62     if name is not None:

InvalidArgumentError:  ValueError: `generator` yielded an element of shape (22, 256, 256, 3) where an element of shape (32, 256, 256, 3) was expected.
Traceback (most recent call last):

  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/ops/script_ops.py", line 243, in __call__
    ret = func(*args)

  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/autograph/impl/api.py", line 309, in wrapper
    return func(*args, **kwargs)

  File "/usr/local/lib/python3.6/dist-packages/tensorflow/python/data/ops/dataset_ops.py", line 821, in generator_py_func
    "of shape %s was expected." % (ret_array.shape, expected_shape))

ValueError: `generator` yielded an element of shape (22, 256, 256, 3) where an element of shape (32, 256, 256, 3) was expected.


	 [[{{node PyFunc}}]]
	 [[IteratorGetNext]] [Op:__inference_train_function_1244]

Function call stack:
train_function


NOTE: I do understand that when we have 3670 training records, the last step(114th) will contain only 22 records and that should not throw an errors. Please correct me if I am wrong.

**Comment by tomerk** [comment]
So two things because It's also come to my attention that apparently the ImageDataGenerator isn't *always* supposed to loop forever.

---------------
1. When you use tf.data `from_generator` and specify a shape, the generator must *always* yield objects of that shape. In your call you're specifying a fixed batch size as part of the shape:
```
ds = tf.data.Dataset.from_generator(lambda: train_data_gen,
output_types=(tf.float32, tf.float32),
output_shapes=([32, 256, 256, 3],
[32, 5])
)
```

When it sees a partial batch of 22 on the last step the shapes do not match so it errors.
You can fix this by leaving the batch size unspecified (as `None`) in `from_generator`:
```
ds = tf.data.Dataset.from_generator(lambda: train_data_gen,
output_types=(tf.float32, tf.float32),
output_shapes=([None, 256, 256, 3],
[None, 5])
)
```
----------------
2. Apparently there is a setting where the ImageDataGenerator isn't supposed to loop forever and shouldn't require steps_per_epoch: If you pass the result of `flow_from_directory` directly to Keras `fit` without converting it to a dataset yourself. In this specific setting the `len` information attached to the ImageDataGenerator sequences has historically been used as an implicit `steps_per_epoch`.

(This does not happen if you manually loop over the generator w/ a for loop, as shown in the documentation of the ImageDataGenerator). This isn't the case for your example code because when you convert it to a dataset manually that cardinality information gets lost.

*However* it seems we did introduce a regression in that setting at some point. (Though we're not positive which exact version of TF it was introduced in). It did make its way into the TF 2.0 release, and we're currently exploring whether to do a patch release to fix it.

**Comment by Thomas-K-John** [comment]
Few points that I found from my testing:
1. As you suggested I tried by leaving the batch size unspecified (as None) in from_generator. This helped me to successfully train the model by successfully running the model.fit() method. But I was not able to use this model for making any predictions. i.e. model.predict() and model.evaluate() methods were just running infinitely again. 

2. I also tried to pass the validation dataset to my fit() method during the training of the model. The model was again running infinitely. Please find the arguments passed to my fit() method below:
model.fit(train_ds, verbose=1,  steps_per_epoch=3670/32, validation_data=validation_ds, epochs =1)

I would like you to please confirm if these two are bugs. Please let me know if I need to raise any seperate bugs of these issues. I would like you to please suggest me any work around to run the model.predict() and model.evaluate() functions.

**Comment by tomerk** [comment]
The steps argument for `evaluate` and `predict` is called `steps`, not `steps_per_epoch`. You could go ahead and set that. For validation the argument is validation_steps.

Alternatively, you could put `.take(num_steps)` at the end of your dataset to force it to have a fixed size.

If we patch the release to handle the case where the generator is directly passed to fit/evaluate/predict it should deal with all of these cases, so you shouldn't need to file any other bugs. We've submitted the code fix already in the TF nightlies, it's just a question of whether it'll go in 2.3 or be patched into 2.2.

**Comment by Thomas-K-John** [comment]
Thank you for the details.

**Comment by driedler** [comment]
The issue appears to be in:
`tensorflow.python.keras.engine.data_adapter.py`  

function:
`select_data_adapter()`

The function is selecting `GeneratorDataAdapter` when it should be selecting `KerasSequenceAdapter`.

If I manually update:
`keras_preprocessing.image.iterator.py`:
```
class Iterator(IteratorType):
```
to:
```
class Iterator(data_utils.Sequence)
```

`flow_from_directory()` works fine with `fit()` again.

**Comment by Thomas-K-John** [comment]
I hope this fix will also resolve the issue faced with predict() and evaluate() methods.

**Comment by tomerk** [comment]
@driedler:
We made a roughly equivalent change to the `keras_preprocessing` pip package, but users have to manually upgrade it to get that effect.

We've also submitted this:
https://github.com/tensorflow/tensorflow/commit/b53ed4d560aaeb7a92185f4fbf2562e5e274456a

which will affect the next version of tensorflow. We're unlikely to patch this into the existing release, so if you want the fix right away go ahead and update keras_preprocessing.

We'll go ahead and close this issue for now.
